ocaml/typing/typedecl.ml

```diff
@@ -1094,6 +1098,54 @@ let check_constraints env sdecl (_, decl) =
       check_constraints_rec env sty.ptyp_loc visited ty
   end
 
+(* CR layouts v3: this is a hack to allow re-exporting the definition
+   of ['a or_null], including constructors, even if one can't define
+   ['a or_null] "honestly", with tools available to users.
+
+   Remove when we allow users to define their own null constructors.
+*)
+let reexport_special_types env loc dpath decl =
+  let or_null_reexport =
+    Builtin_attributes.has_or_null_reexport decl.type_attributes
+  in
+  match or_null_reexport with
+  | false -> decl
+  | true ->
+    let fail () =
+      raise (Error (loc, Invalid_reexport
+        { definition = dpath; expected = Predef.path_or_null }))
+    in
+    match decl.type_kind, decl.type_manifest with
+    (* We require users to define ['a t = 'a or_null]. By assumption,
+       they can't define the valid type kind, so it must be left abstract.
+       Manifest must be set to [or_null] so typechecking stays correct. *)
+    | Type_abstract _, Some ty ->
+      let param =
+        match get_desc ty with
+        | Tconstr(path, [param], _)
+          when Path.same path Predef.path_or_null -> param
+        | _ -> fail ()
+      in
+      let or_null_decl = Env.find_type Predef.path_or_null env in
```

Comment:
Maybe I'm wrong, but it seems easier to me just to construct the right kind from scratch instead of taking `or_null`'s kind and altering it. Make a little function do this and exporting it from predef will avoid code duplication.